### PR TITLE
feat: validate airbyte metadata added to iceberg schema

### PR DESCRIPTION
**Pull request by jdpgrailsdev**
## What
* Validate that Airbyte metadata columns are added to the Iceberg schema
* Actually detect insert vs update operation

## How
* Add tests to verify that Airbyte metadata columns are added to the Iceberg schema
* Enable CDK tests to validate metadata is added to a schema when `withMetadata()` is called
* Convert `IcebergUtil` to a Micronaut singleton to better facilitate testing
* Detect insert/update when converting to Iceberg record

## Review guide
1. `IcebergUtil.kt`
2. `IcebergUtilTest.kt`
3. `IcebergV2WriterTest.kt`
4. `AirbyteTypeToAirbyteTypeWithMetaTest.kt`

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by edgao** [comment]
this looks a lot like https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteTypeToAirbyteTypeWithMeta.kt ? which we can maybe reuse in some way

File: airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/AirbyteTypeToIcebergSchema.kt [hunk 1]
```diff
@@ -90,16 +76,90 @@ class AirbyteTypeToIcebergSchema {
...
+            objectType.properties.entries.map { (name, field) ->
+                if (field.nullable) {
+                    NestedField.optional(UUID.randomUUID().hashCode(), name, convert(field.type))
+                } else {
+                    NestedField.required(UUID.randomUUID().hashCode(), name, convert(field.type))
+                }
+            }
+    }
+
+    private fun createAirbyteMetadataFields(): List<NestedField> {
```
Comment (subodh1810) [inline airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/AirbyteTypeToIcebergSchema.kt:91]: Most of these columns are being added to the schema using the mapper architecture. if you take a look at 
```
val schema = pipeline.finalSchema.withAirbyteMeta(true).toIcebergSchema(primaryKeys)
```
In IcebergV2Writer. The `withAirbyteMeta` method is adding a bunch of em. If we want to add more like the generation id , etc. we should use the same design.

File: airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt [hunk 2]
```diff
@@ -185,4 +188,15 @@ object IcebergUtil {
         schema.identifierFieldNames().forEach { builder.asc(it) }
         return builder.build()
     }
+
+    private fun getOperation(record: DestinationRecord, schema: Schema): Operation =
+        if (record.data.toJson().get(AIRBYTE_CDC_DELETE_COLUMN) != null) {
```
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt:193]: I wonder how costly this `toJson` method call on each record would be. Do you think we should do this upstream when the destination receives an actual json from the platform?

Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt:193]: @subodh1810 I was concerned about this too, but I don't think that there is any other way at the moment to discover this.  We can comment this out for now/not support deletes at the moment.  Ideally, we would add this information to the protocol so we can test it directly without serializing the payload.

**Comment by jdpgrailsdev** [comment]
> this looks a lot like https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteTypeToAirbyteTypeWithMeta.kt ? which we can maybe reuse in some way

@edgao It does and I wasn't sure how to reuse that given that it returns Airbyte types and not Iceberg types.  I suspect what is really being pointed out here is some refactor to this in the CDK to make the common parts more common.

File: airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt (hunk 2 above)
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt]: I have disabled the delete check for now and added comments about why to the code

File: airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt [hunk 3]
```diff
@@ -180,9 +188,36 @@ object IcebergUtil {
...
+    private fun getOperation(
+        @Suppress("UNUSED_PARAMETER") record: DestinationRecord,
+        importType: ImportType,
+    ): Operation =
+        // TODO This disabled out of caution due to the potential cost of serializing every record
+        // to check
+        //        the deletion status.  This should be revisited when there is a cheaper way to do
+        // this, such
+        //        as after a protocol change that explicitly states the operation.
+        //        if (record.data.toJson().get(AIRBYTE_CDC_DELETE_COLUMN) != null) {
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt:215]: can we do something like this to dodge the serialize?
```
(record.data as ObjectValue).properties[DELETE_COLUMN] != null
```

**Review by edgao — APPROVED** [review]
lgtm, just a small question about the deleted_at detection

File: airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt (hunk 3 above)
Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt:215]: @edgao Sure?  I'll test it out and update the PR if it works

Comment (jdpgrailsdev) [reply airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/IcebergUtil.kt:215]: @edgao That appeared to work.  I will update the PR to include the DELETE operation detection based on your suggestion.

**Pull request merged**